File: backend/api/repository.py
"""
仓库操作API端点
"""

from fastapi import APIRouter, HTTPException

from models.schemas import (
    ApiResponse,
    BranchesResponse,
    BranchInfo,
    CheckoutCommitRequest,
    CommitInfo,
    CreateBranchRequest,
    CreateCommitRequest,
    FileChange,
    InitRepositoryRequest,
    MergeBranchRequest,
    RepositoryStatus,
    SwitchBranchRequest,
)
from services.git_wrapper import (
    GitError,
    GitWrapper,
    InvalidPathError,
    MergeConflictError,
    RepositoryNotFoundError,
)

router = APIRouter(prefix="/repository", tags=["repository"])


@router.post("/init", response_model=ApiResponse)
async def init_repository(request: InitRepositoryRequest):
    """
    初始化Git仓库

    创建新的Git仓库并生成.chronos配置文件
    """
    try:
        wrapper = GitWrapper(request.path)
        result = wrapper.init_repository()

        return ApiResponse(
            success=result["success"],
            message=result["message"],
            data={"already_initialized": result.get("already_initialized", False)},
        )
    except InvalidPathError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.get("/status", response_model=ApiResponse)
async def get_status(path: str):
    """
    获取仓库状态

    返回当前分支和文件变更列表
    """
    try:
        wrapper = GitWrapper(path)
        status = wrapper.get_status()

        # 转换为Pydantic模型
        changes = [FileChange(**change) for change in status["changes"]]
        status_data = RepositoryStatus(
            branch=status["branch"], changes=changes, is_clean=status["is_clean"]
        )

        return ApiResponse(
            success=True, message="获取状态成功", data=status_data.model_dump()
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.post("/commit", response_model=ApiResponse)
async def create_commit(request: CreateCommitRequest):
    """
    创建快照（提交）

    将变更保存为新的提交
    """
    try:
        wrapper = GitWrapper(request.path)
        result = wrapper.create_commit(request.message, request.files)

        if not result["success"]:
            return ApiResponse(success=False, message=result["message"])

        # 转换提交信息
        commit_data = None
        if "commit" in result and result["commit"]:
            commit_data = CommitInfo(**result["commit"]).model_dump()

        return ApiResponse(
            success=True, message=result["message"], data={"commit": commit_data}
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.get("/log", response_model=ApiResponse)
async def get_log(path: str, limit: int = None, branch: str = None):
    """
    获取提交历史

    返回提交历史记录列表
    """
    try:
        wrapper = GitWrapper(path)
        commits = wrapper.get_log(limit=limit, branch=branch)

        # 转换为Pydantic模型
        commits_data = [CommitInfo(**commit).model_dump() for commit in commits]

        return ApiResponse(
            success=True,
            message=f"获取到{len(commits_data)}条提交记录",
            data={"logs": commits_data},
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.post("/checkout", response_model=ApiResponse)
async def checkout_commit(request: CheckoutCommitRequest):
    """
    回滚到指定提交

    将工作区恢复到指定提交的状态
    """
    try:
        wrapper = GitWrapper(request.path)
        result = wrapper.checkout_commit(request.commit_id)

        return ApiResponse(
            success=result["success"],
            message=result["message"],
            data={
                "commit_id": result.get("commit_id"),
                "has_changes": result.get("has_changes", False),
            },
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.get("/branches", response_model=ApiResponse)
async def get_branches(path: str):
    """
    获取分支列表

    返回所有分支和当前分支
    """
    try:
        wrapper = GitWrapper(path)
        result = wrapper.get_branches()

        # 转换为Pydantic模型
        branches = [BranchInfo(**branch) for branch in result["branches"]]
        branches_data = BranchesResponse(branches=branches, current=result["current"])

        return ApiResponse(
            success=True, message="获取分支列表成功", data=branches_data.model_dump()
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.post("/branch", response_model=ApiResponse)
async def create_branch(request: CreateBranchRequest):
    """
    创建新分支

    在当前提交创建新分支
    """
    try:
        wrapper = GitWrapper(request.path)
        result = wrapper.create_branch(request.branch_name)

        return ApiResponse(
            success=result["success"],
            message=result["message"],
            data={"branch": result.get("branch")},
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.post("/switch", response_model=ApiResponse)
async def switch_branch(request: SwitchBranchRequest):
    """
    切换分支

    切换到指定分支
    """
    try:
        wrapper = GitWrapper(request.path)
        result = wrapper.switch_branch(request.branch_name)

        return ApiResponse(
            success=result["success"],
            message=result["message"],
            data={
                "branch": result.get("branch"),
                "has_changes": result.get("has_changes", False),
            },
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")


@router.post("/merge", response_model=ApiResponse)
async def merge_branch(request: MergeBranchRequest):
    """
    合并分支

    将源分支合并到目标分支
    """
    try:
        wrapper = GitWrapper(request.path)
        result = wrapper.merge_branch(request.source_branch, request.target_branch)

        return ApiResponse(
            success=result["success"],
            message=result["message"],
            data={"source": result.get("source"), "target": result.get("target")},
        )
    except RepositoryNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except MergeConflictError as e:
        raise HTTPException(
            status_code=409, detail={"message": str(e), "conflicts": e.conflicts}
        )
    except GitError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"未知错误: {str(e)}")

File: src/server/routes/repository.ts
/**
 * Chronos v2.0 - Repository Routes
 * 
 * 仓库相关的 API 路由
 */

import { Router } from 'express'
import { GitService } from '../services/git-service'
import { FileService } from '../services/file-service'
import { DatabaseService } from '../services/database-service'
import { createError, isAppError } from '../utils/errors'
import { logger } from '../utils/logger'
import { generateId } from '@/shared/utils'
import { ERROR_CODES } from '@/shared/constants'
import type { ApiResponse } from '@/shared/types'
import path from 'path'

const router = Router()
const gitService = new GitService()
const fileService = new FileService()

// 数据库服务（需要在应用启动时初始化）
let dbService: DatabaseService

export function initializeDatabase(db: DatabaseService): void {
  dbService = db
}

/**
 * POST /api/repository/init
 * 初始化仓库
 */
router.post('/init', async (req, res) => {
  try {
    const { path: repoPath } = req.body

    if (!repoPath) {
      throw createError.invalidPath('路径不能为空')
    }

    // 检查路径是否存在
    const exists = await fileService.exists(repoPath)
    if (!exists) {
      throw createError.repoNotFound(repoPath)
    }

    // 检查是否为目录
    const isDir = await fileService.isDirectory(repoPath)
    if (!isDir) {
      throw createError.invalidPath('路径必须是文件夹')
    }

    // 初始化 Git 仓库
    await gitService.init(repoPath)

    // 保存到数据库
    const repoId = generateId()
    dbService.saveRepository({
      id: repoId,
      path: repoPath,
      name: path.basename(repoPath),
      currentBranch: 'main',
      lastOpened: new Date(),
    })

    dbService.recordRepositoryOpen(repoId)

    logger.info('Repository initialized', { path: repoPath })

    const response: ApiResponse = {
      success: true,
      data: {
        id: repoId,
        path: repoPath,
        name: path.basename(repoPath),
      },
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to initialize repository', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * POST /api/repository/open
 * 打开仓库
 */
router.post('/open', async (req, res) => {
  try {
    const { path: repoPath } = req.body

    if (!repoPath) {
      throw createError.invalidPath('路径不能为空')
    }

    // 检查路径是否存在
    const exists = await fileService.exists(repoPath)
    if (!exists) {
      throw createError.repoNotFound(repoPath)
    }

    // 检查是否已初始化
    const isInit = await gitService.isInitialized(repoPath)
    if (!isInit) {
      throw createError.repoNotInitialized(repoPath)
    }

    // 从数据库获取或创建仓库记录
    let repo = dbService.getRepositoryByPath(repoPath)

    if (!repo) {
      const repoId = generateId()
      dbService.saveRepository({
        id: repoId,
        path: repoPath,
        name: path.basename(repoPath),
        currentBranch: 'main',
        lastOpened: new Date(),
      })
      repo = dbService.getRepository(repoId)
    } else {
      // 更新最后打开时间
      dbService.saveRepository({
        ...repo,
        lastOpened: new Date(),
      })
    }

    if (repo) {
      dbService.recordRepositoryOpen(repo.id)
    }

    logger.info('Repository opened', { path: repoPath })

    const response: ApiResponse = {
      success: true,
      data: repo,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to open repository', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * GET /api/repository/status
 * 获取仓库状态
 */
router.get('/status', async (req, res) => {
  try {
    const { path: repoPath } = req.query

    if (!repoPath || typeof repoPath !== 'string') {
      throw createError.invalidPath('路径不能为空')
    }

    // 检查路径是否存在
    const exists = await fileService.exists(repoPath)
    if (!exists) {
      throw createError.repoNotFound(repoPath)
    }

    const status = await gitService.getStatus(repoPath)

    const response: ApiResponse = {
      success: true,
      data: status,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to get repository status', error)

    if (isAppError(error)) {
      // 根据错误码返回不同的HTTP状态码
      const statusCode = error.code === ERROR_CODES.REPO_NOT_FOUND ? 404 : 400
      res.status(statusCode).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * 递归地将状态信息添加到文件树中
 */
function addStatusToFileTree(files: any[], statusMap: Map<string, string>): any[] {
  return files.map(file => {
    const fileWithStatus = {
      ...file,
      status: statusMap.get(file.path)
    }
    
    if (file.children && file.children.length > 0) {
      fileWithStatus.children = addStatusToFileTree(file.children, statusMap)
    }
    
    return fileWithStatus
  })
}

/**
 * GET /api/repository/files
 * 获取文件列表（包含状态信息）
 */
router.get('/files', async (req, res) => {
  try {
    const { path: repoPath } = req.query

    if (!repoPath || typeof repoPath !== 'string') {
      throw createError.invalidPath('路径不能为空')
    }

    // 获取文件列表和状态
    const [files, status] = await Promise.all([
      fileService.scanDirectory(repoPath),
      gitService.getStatus(repoPath)
    ])

    // 创建状态映射表
    const statusMap = new Map(
      status.changes.map(change => [change.path, change.status])
    )

    // 递归地将状态信息合并到文件树中
    const filesWithStatus = addStatusToFileTree(files, statusMap)

    const response: ApiResponse = {
      success: true,
      data: { files: filesWithStatus },
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to get files', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * POST /api/repository/commit
 * 创建快照
 */
router.post('/commit', async (req, res) => {
  try {
    const { path: repoPath, message, files } = req.body

    if (!repoPath) {
      throw createError.invalidPath('路径不能为空')
    }

    if (!message) {
      throw createError.gitError('提交信息不能为空')
    }

    const commitId = await gitService.createCommit(repoPath, message, files)

    logger.info('Commit created', { path: repoPath, commitId })

    const response: ApiResponse = {
      success: true,
      data: { commitId },
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to create commit', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * GET /api/repository/log
 * 获取历史记录
 */
router.get('/log', async (req, res) => {
  try {
    const { path: repoPath, limit } = req.query

    if (!repoPath || typeof repoPath !== 'string') {
      throw createError.invalidPath('路径不能为空')
    }

    const limitNum = limit ? parseInt(limit as string, 10) : undefined

    const logs = await gitService.getLog(repoPath, limitNum)

    const response: ApiResponse = {
      success: true,
      data: { logs },
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to get log', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * POST /api/repository/checkout
 * 回滚到指定快照
 */
router.post('/checkout', async (req, res) => {
  try {
    const { path: repoPath, commitId } = req.body

    if (!repoPath) {
      throw createError.invalidPath('路径不能为空')
    }

    if (!commitId) {
      throw createError.gitError('提交 ID 不能为空')
    }

    await gitService.checkout(repoPath, commitId)

    logger.info('Checked out commit', { path: repoPath, commitId })

    const response: ApiResponse = {
      success: true,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to checkout', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * GET /api/repository/branches
 * 获取分支列表
 */
router.get('/branches', async (req, res) => {
  try {
    const { path: repoPath } = req.query

    if (!repoPath || typeof repoPath !== 'string') {
      throw createError.invalidPath('路径不能为空')
    }

    const branches = await gitService.getBranches(repoPath)

    const response: ApiResponse = {
      success: true,
      data: { branches },
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to get branches', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * POST /api/repository/branch
 * 创建分支
 */
router.post('/branch', async (req, res) => {
  try {
    const { path: repoPath, branchName } = req.body

    if (!repoPath) {
      throw createError.invalidPath('路径不能为空')
    }

    if (!branchName) {
      throw createError.gitError('分支名称不能为空')
    }

    await gitService.createBranch(repoPath, branchName)

    logger.info('Branch created', { path: repoPath, branchName })

    const response: ApiResponse = {
      success: true,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to create branch', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * POST /api/repository/switch
 * 切换分支
 */
router.post('/switch', async (req, res) => {
  try {
    const { path: repoPath, branchName } = req.body

    if (!repoPath) {
      throw createError.invalidPath('路径不能为空')
    }

    if (!branchName) {
      throw createError.gitError('分支名称不能为空')
    }

    await gitService.switchBranch(repoPath, branchName)

    logger.info('Switched branch', { path: repoPath, branchName })

    const response: ApiResponse = {
      success: true,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to switch branch', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * POST /api/repository/merge
 * 合并分支
 */
router.post('/merge', async (req, res) => {
  try {
    const { path: repoPath, sourceBranch, targetBranch } = req.body

    if (!repoPath) {
      throw createError.invalidPath('路径不能为空')
    }

    if (!sourceBranch) {
      throw createError.gitError('源分支不能为空')
    }

    const result = await gitService.mergeBranch(repoPath, sourceBranch, targetBranch)

    logger.info('Merged branch', { path: repoPath, sourceBranch, targetBranch })

    const response: ApiResponse = {
      success: true,
      data: result,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to merge branch', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * GET /api/repository/recent
 * 获取最近使用的仓库列表
 */
router.get('/recent', async (req, res) => {
  try {
    const { limit } = req.query
    const limitNum = limit ? parseInt(limit as string, 10) : 10

    const repositories = dbService.getRecentRepositories(limitNum)

    // 验证路径是否仍然存在
    const validRepositories = []
    for (const repo of repositories) {
      const exists = await fileService.exists(repo.path)
      if (exists) {
        validRepositories.push(repo)
      }
    }

    logger.info('Retrieved recent repositories', { count: validRepositories.length })

    const response: ApiResponse = {
      success: true,
      data: { repositories: validRepositories },
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to get recent repositories', error)

    res.status(500).json({
      success: false,
      error: '服务器错误',
    })
  }
})

/**
 * DELETE /api/repository/:id
 * 删除仓库记录（软删除 - 不删除实际文件）
 */
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params

    if (!id) {
      throw createError.invalidPath('仓库 ID 不能为空')
    }

    dbService.deleteRepository(id)

    logger.info('Repository record deleted (soft delete)', { id })

    const response: ApiResponse = {
      success: true,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to delete repository', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

/**
 * POST /api/repository/destroy
 * 完全移除时光机管理（硬删除 - 删除 .git 文件夹）
 */
router.post('/destroy', async (req, res) => {
  try {
    const { id } = req.body

    if (!id) {
      throw createError.invalidPath('仓库 ID 不能为空')
    }

    // 获取仓库信息
    const repo = dbService.getRepository(id)
    if (!repo) {
      throw createError.repoNotFound('仓库不存在')
    }

    // 先删除 .gitignore 文件（在删除 .git 之前）
    const gitignorePath = path.join(repo.path, '.gitignore')
    const gitignoreExists = await fileService.exists(gitignorePath)

    if (gitignoreExists) {
      await import('fs/promises').then(fs => fs.unlink(gitignorePath))
      logger.info('Removed .gitignore file', { path: gitignorePath })
    }

    // 再删除 .git 文件夹
    const gitPath = path.join(repo.path, '.git')
    const gitExists = await fileService.exists(gitPath)

    if (gitExists) {
      await import('fs/promises').then(fs => fs.rm(gitPath, { recursive: true, force: true }))
      logger.info('Removed .git directory', { path: gitPath })
    }

    // 删除数据库记录
    dbService.deleteRepository(id)

    logger.info('Repository completely destroyed (hard delete)', { id, path: repo.path })

    const response: ApiResponse = {
      success: true,
    }

    res.json(response)
  } catch (error) {
    logger.error('Failed to destroy repository', error)

    if (isAppError(error)) {
      res.status(400).json({
        success: false,
        error: error.message,
        code: error.code,
      })
    } else {
      res.status(500).json({
        success: false,
        error: '服务器错误',
      })
    }
  }
})

export default router
